File: src/main/java/org/flossware/civilization/model/WorldConstraints.java
package org.flossware.civilization.model;

/**
 * Global world parameters affecting all civilizations.
 */
public record WorldConstraints(
    double politicalStability,
    double warFrequency,
    double climateVolatility,
    double plagueProbability,
    double resourceAbundance,
    double randomEventFrequency
) {
    public WorldConstraints {
        validateRange(politicalStability, "politicalStability");
        validateRange(warFrequency, "warFrequency");
        validateRange(climateVolatility, "climateVolatility");
        validateRange(plagueProbability, "plagueProbability");
        validateRange(resourceAbundance, "resourceAbundance");
        validateFrequencyRange(randomEventFrequency, "randomEventFrequency");
    }

    private void validateRange(double value, String name) {
        if (value < 0 || value > 1) {
            throw new IllegalArgumentException(name + " must be in [0, 1]: " + value);
        }
    }

    private void validateFrequencyRange(double value, String name) {
        if (value < 0 || value > 2) {
            throw new IllegalArgumentException(name + " must be in [0, 2]: " + value);
        }
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/WorldConstraints.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable

@Serializable
data class WorldConstraints(
    val politicalStability: Double,
    val warFrequency: Double,
    val climateVolatility: Double,
    val plagueProbability: Double,
    val resourceAbundance: Double
) {
    init {
        require(politicalStability in 0.0..1.0) { "politicalStability must be in [0, 1]" }
        require(warFrequency in 0.0..1.0) { "warFrequency must be in [0, 1]" }
        require(climateVolatility in 0.0..1.0) { "climateVolatility must be in [0, 1]" }
        require(plagueProbability in 0.0..1.0) { "plagueProbability must be in [0, 1]" }
        require(resourceAbundance in 0.0..1.0) { "resourceAbundance must be in [0, 1]" }
    }
}
